src/stim/util_top/reference_sample_tree.cc

```diff
@@ -186,6 +186,32 @@ bool ReferenceSampleTree::operator!=(const ReferenceSampleTree &other) const {
     return !(*this == other);
 }
 
+bool ReferenceSampleTree::operator[](size_t index) const {
+    size_t current_absolute_index = 0;
+    bool result;
+    bool value_found = try_get_bit_value(index, current_absolute_index, result);
+    assert(value_found);
+    return result;
+}
+
+bool ReferenceSampleTree::try_get_bit_value(
+    size_t desired_absolute_index, size_t &current_absolute_index, bool &bit_value) const {
+    for (uint64_t k = 0; k < repetitions; k++) {
```

Comment:
I may be able to do this is max 2 iterations, instead of `repetitions` iterations. Will update with new commit soon.

Reply:
Yeah, if you compute the size of the elements then you can skip ahead to the correct iteration with a division. That will be very important for getting ~O(1) time in practice rather than ~O(n).

Create a test case that encodes a tree with a quadrillion bits (but doesn't decompress it) and tries to index into it. Passing corresponds to it taking less than a milliescond (rather than hours).

Reply:
> if you compute the size of the elements then you can skip ahead to the correct iteration with a division.

Yes, but if you actually used `size()` it would always require a full iteration of the tree, even for the first element.
What I changed this to is to do 1 iteration first, which both looks for the element at the desired index (if it happens to be there) and also calculates the size. Then, if the element was not found, we skip forward and (if still in the node) do a second pass to actually find the element. Thus "max 2 iterations" (how that many iterations it actually is in the full recursive tree is a bit more complicated).
I don't think this can be faster without storing the node size as an element (so it's precomputed).

Reply:
> Create a test case that encodes a tree with a quadrillion bits (but doesn't decompress it) and tries to index into it.

Did a test with >10^17 bits (encoded into 5 nodes).
Tested a handful of bits from the beginning, the end, and 1 outer repetition before then ending (~1 million bits before the end).
Bazel shows it as taking less than 1ms.